### ISSUE TITLE
support web GUI interface for proxy rama bin cli

### ISSUE THREAD

**Issue by GlenDC** [plabayo/rama#11]
`rama proxy ...`

**Comment by GlenDC** [comment]
As I am new to writing TUIs, I think I'll start with reading through https://ratatui.rs/introduction/ first and follow those tutorials.

**Comment by pm100** [comment]
I know [rata]tui well, have contributed to several projects using it.
What features were you thinking that a tui for the cli should have

**Comment by GlenDC** [comment]
The title is a bit misleading though. To put some context, rama is above all a framework and the strength is in "code as proxy", or however you want to put it. There are already plenty of well established and more feature rich proxy binaries out there. Rama's strength is mostly in the fact that it's a modular framework that you can easily mix and match with custom layers, services and other code on any layer of the proxy stack (or server / client in case you are not building a proxy).

It's an old issue and one that has a pretty big scope. If you are interested in contributing for this, which I would welcome very much, we probably best split this up in sub issues to make it easier to review but also nicer for you to get it merged in pieces instead of having it most likely dragged on for longer than any party wishes.

The end goal of this story is that one can easily expose a graphical interface for their client/server/proxy, very similar to something like `mitmproxy` / `tshark`. As in that we can group requests per connection, see the details, copy them as curl, redo requests with or without modifications ("repeat") etc... Sky is the limit, as such best we should do it step by step I imagine.

That end goal is achieved by having:

- services/layers ready that one can easily plugin to their own rama-based project
- support using those services/layers added to rama-cli for demo case

(rama-cli is for now mostly a minimal demo-case, as I said, for the time being until we have more features and rich config support there is not that much strength in the cli compared to much more established options out there)

Goal would also be that we can decouple the "rendering" / "graphical" part from the actual implementation. Such that you could expose it as a TUI, but also for example as a native application or web GUI. Ideally the only unique code to that is the code to render and communicate, but that the data collection/feed aspect of it is common between these.

I don't think that I did a good job at explaining it as-is, but despite my poor attempt, do you more or less understand where I think it should go? What is your opinion on the matter? How do you see it. Is that something you are interested in? if so, what are the steps that you have in mind yourself?

**Comment by pm100** [comment]
> I don't think that I did a good job at explaining it as-is, but despite my poor attempt, do you more or less understand where I think it should go? What is your opinion on the matter? How do you see it. Is that something you are interested in? if so, what are the steps that you have in mind yourself?


I understand that rama is a toolkit and that the cli is 'just' a demo of a couple of simple examples. This is why I asked what would make sense . I mean it could be
- select which service to run
- configure the service
- monitor its state, clients, message count....
- display the log

But is that really useful.

What seems like it would be useful would be a tool that can peek inside the service stack of a running 'app' , but surely thats the job of open telemetry (I dont know that at all , just guessing based on the name)

**Comment by GlenDC** [comment]
Open telemetry is for distilled metric information, not really the full detail and scope that something like this would be able to handle. As that wouldn't really scale. This is really more of a developer thing on low scale.

You can play with https://mitmproxy.org/ if you want to know what I mean.

here are some screenshots specifically:

request list:

<img width="1476" alt="Image" src="https://github.com/user-attachments/assets/9e9424fb-6ad8-4680-b061-ce3b15622b8f" />

request detail:

<img width="1475" alt="Image" src="https://github.com/user-attachments/assets/4ab5afe6-15b9-400a-8bb1-ec25d8181789" />

The way you would achieve that is by having some kind of producer and consumer concept.
In it's most simplest form it would be a layer that you use on the http layer to extract information from the request and response as well as the context. It would then send it to a consumer which consumes these details. To make it more efficient it would be a layer for the transport layer (with or without tls) and a layer from http separately, such that information for the connection specifically, tagging and tls would be done only once and not for each request.

That information could be send to a runtime that handles all that logic. E.g a tui client, but someone else can just as well make a fat application of it or a web GUI or w/e.

Does that make sense?

**Comment by GlenDC** [comment]
Implementing this one now myself, but going to do it via web gui... one can contribute a tui in future, we already ahve ratatui as dep, but honestly I feel something like this makes a lot more sense in a web browser, and thanks to our html/datastar support now it's a pretty easy job.
